### ISSUE TITLE
Difference in derive(Deref) from derive_deref

### ISSUE THREAD

**Issue by spease** [JelteF/derive_more#48]
Looks like this crate derives deref based on the enclosed item's implementation of deref, whereas derive_deref sets up deref to deref to the enclosed item.

EG where
```
B::deref() -> &C

#[derive(Deref)]
struct A(B);
```

derive_deref
```
A::deref -> &B
```

derive_more
```
A::deref -> &C
```

For the code I've worked with, the former case is usually the more commonly desired one since I've often had to use wrapper structs for serialization that should otherwise be transparent.

I'm not sure if that's the absolute most common desired implementation though.

**Comment by JelteF** [comment]
I understand that your usecase makes sense, but I don't really think it fits well in this library. For derive `Deref` usecase I actually like the more explicit derive `Into` that this lib provides better for this behaviour. However, I don't think you could also use it as an equivalent for `DerefMut`.

I hadn't seen derive_deref before though btw. It does seem to work both for my current implementation, dereferencing twice when the inner type implements `Deref` as well, but having the other behaviour otherwise. I had seen [shrinkwraprs]( https://crates.io/crates/shrinkwraprs) before though, which does sort of the same, but derives some more traits and can also use some attributes to choose field to deref. 
Because that one does `derive(Shrinkwrap)` you can also use that sometimes and still use my implementation for the cases where you want to deref twice.

I will leave this issue open for now, because the docs for `Deref` and `DerefMut` should contain this information as well.

**Comment by spease** [comment]
So the recommended solution here is to use shrinkwrap for those situations where the struct is a transparent wrapper sort of object?

General cases where this comes up is for custom serialization, and forcing type safety although the data storage is still a string, u64, etc.

**Comment by JelteF** [comment]
My advice would be:
1. use `Into` if you don't need mutable access
2. derive `Deref` and `DerefMut` if what you wrap is a pointer type, e.g `&`, `Rc` or `Box` 
3. Otherwise use shrinkwraprs

**Comment by jsdw** [comment]
I think I just bumped into the same issue (using 0.12.0); I was surprised to find that

```
#[derive(Deref)]
struct Wrapper(u64);
```

leads to the error:

```
error[E0277]: the trait bound `u64: std::ops::Deref` is not satisfied
  --> src/main.rs:21:10
   |
21 | #[derive(Deref)]
   |          ^^^^^ the trait `std::ops::Deref` is not implemented for `u64`
```

My belief was that the idiomatic approach to deriving Deref is to unwrap one level. The compiler will insert as many Derefs as needed when passing references to things into functions I think, so unwrapping one level here feels like the more flexible approach.

Thanks for your crate :)

**Comment by jesskfullwood** [comment]
Just want to note that I was caught out by this too. Maybe it could be added as a separate macro? (`DerefOnce`? `NewtypeDeref`?)

**Comment by Nokel81** [comment]
I completely agree with @jsdw here, the compiler already adds as many dereferences as necessary so there is really no need to have this crate try and dereference twice.

**Comment by magnet** [comment]
We're now using `derive_deref` *and* `derive_more` because of this limitation. I also think it would be backwards compatible and more ergonomic to adapt `derive_deref`'s behavior.

**Comment by JelteF** [comment]
I've thought about this some more and I agree that the proposed behaviour would be better. If especially because of what @jsdw wrote. I don't have the time to do this myself in the near future though, but I do welcome a PR for the change.

**Comment by magnet** [comment]
Cool, I will submit a PR.

**Comment by kgv** [comment]
Is there any progress?

**Comment by magnet** [comment]
Hey, I built the PR and never submitted it. I'll submit it later today.

**Comment by JelteF** [comment]
On master `Deref` is now implemented like this issue wants. To get the old behaviour you can do:
```rust
#[derive(Deref)]
#[deref(forward)]
struct A(Box<i32>);

**Comment by JelteF** [comment]
This is now released as version 0.99.0. Announcement link: https://reddit.com/r/rust/comments/duptc0/derive_more_0990_released_biggest_release_ever
